src/basic/escape.c

```diff
@@ -510,7 +510,7 @@ char* shell_maybe_quote(const char *s, EscapeFlags flags) {
                     strchr(SHELL_NEED_QUOTES, *p))
                         break;
 
-        if (!*p)
+        if (!*p && (p > s || !FLAGS_SET(flags, ESCAPE_EMPTY)))
                 return strdup(s);
```

Comment:
i think it would be a lot more readable to place a separate check at the top of the function:

```c
if (FLAGS_SET(flags, ESCAPE_EMPTY) && isempty(s))
        return strdup("\"\"");
```

i.e. this is a special case, lets give it a special codepath...